File: Python/update_grid.py
import os # to access credentials stored as environment variables and create local cache directory if needed
import requests # for HTML Post and Get methods
import json # To parse response payloads
import pandas as pd # For consolidating results into a dataframe
import zipfile

def ensure_local_folder(path_folder):
    ''' 
    To check for existence of named local folder and create folder if absent
    
    Inputs ---
    path_folder   str   name of folder required
    
    Outputs ---
    None
    ''' 
    
    # Test inputs
    if type(path_folder) is not str:
        raise ValueError('path_folder supplied to ensure_local_folder is not string')
    
    # Test for path folder
    if not os.path.isdir(path_folder):
        print('Creating local folder:\n ' + path_folder + '\nIn:\n ' + os.getcwd())
        os.mkdir(path_folder)


def update_grid(path_data = "./data/"):
    ensure_local_folder(path_data)
    
    # Check release details
    res_details = requests.get("https://api.figshare.com/v2/collections/3812929/articles")
    if res_details.status_code != 200:
        value_error_msg = ''.join(('\nRequest to update_grid has failed to retrieve release details from figshare with status:\n',
                                   str(res_details.status_code)))
        raise ValueError(value_error_msg)

    details_df = pd.DataFrame(json.loads(res_details.text))
    max_date = details_df["published_date"].max()
    release_df = details_df.loc[details_df["published_date"]==max_date, ['url', 'published_date']]

    # Get most recent filename and download link
    res_files = requests.get(release_df['url'][0])
    if res_files.status_code != 200:
        value_error_msg = ''.join(('\nRequest to update_grid has failed to retrieve file details from figshare with status:\n',
                                   str(res_files.status_code)))
        raise ValueError(value_error_msg)
    res_files = json.loads(res_files.text).get('files')[0]

    # Download release if file does not exist in path_data
    if os.path.exists(path_data + "download_log.txt"):
        if res_files.get('name') == open(path_data + 'download_log.txt', 'r').read().split('\n', 1)[0]:
            print("Local GRID data cache already latest release")

            # Exit update_grid without further side effects
            return

    # Download release 
    print("Downloading latest GRID release ("+ max_date + ") to " + path_data)
    res = requests.get(res_files.get('download_url'), allow_redirects=True)
    open(path_data + res_files.get('name'), 'wb').write(res.content)

    # Unzip release
    with zipfile.ZipFile(path_data + res_files.get('name'), 'r') as zip_ref:
        zip_ref.extractall(path_data)

    # Log release via release file name
    open(path_data + "download_log.txt", 'w').write(res_files.get('name') + '\n')
    
    # Tidy up
    if os.path.exists(path_data + res_files.get('name')):
        os.remove(path_data + res_files.get('name'))


if __name__ == "__main__":
    # execute only if run as a script
    main()

File: R/update_grid.R
#update_grid.R
#
#Seferin James, BA, MLitt, PhD
#
#2021/05/04
#
#grid_update is an R script that checks for the latest release of the GRID database of institutional data by Digital Science (https://www.grid.ac/). Details about the most recent update are retrieved via the figshare api. The most recent data is downloaded as a zip file if it is not already available locally and extracted. A log file is created and the zip file tidied up.

#Scrip depends on base, curl and jsonlite packages.

update_grid <- function (path_data = "./data/") {
  required_packages <- c("curl", "jsonlite")
  lapply(required_packages, require, character.only = TRUE)
  
  #Check for valid path_data value
  if(!is.na(path_data)){
    if(path_data != ""){
      
      if(!grepl("/$", path_data)){
        path_data <- paste0(path_data, "/")
      }
      
    }else{
      warning("path_data is NULL in grid_update call")
      return(NULL)
    }
  }else{
    
    warning("path_data is NA in grid_update call")
    return(NULL)
    
  }
  
  #Ensure path_data exists
  if(!dir.exists(path_data)){
    dir.create(path_data)
    if(!dir.exists(path_data)){
      warning("path_data does not exist and an attempt to create it has failed")
      return(NULL)
    }
  }
  
  #Function to get information on GRID releases, returns url and date of most recent release
  #The two web get functions could possibly be consoidated into one that returns json res
  web_get_update_details <- function () {
    res <- curl_fetch_memory("https://api.figshare.com/v2/collections/3812929/articles")
    res <- fromJSON(rawToChar(res$content))[,c("url", "published_date")]
    res <- res[which(res$published_date == max(res$published_date)),]
    return(res)
    }
  
  #Function to retrieve the most recent filename and download link
  web_get_file_details <- function (url) {
    res <- curl_fetch_memory(url)
    res <- fromJSON(rawToChar(res$content))
    res <- as.data.frame(res[['files']])
    res <- res[,c("name","download_url")]
    return(res)
  }
  
  web_get_file_download <- function(file_details, path_data){
    download.file(file_details$download_url, 
                  paste0(path_data, file_details$name), mode="wb")
    
    #Strip trailing slash from path_data prior to unzip method
    safe_path_data <- gsub("/$", "", path_data)
    
    #Unzip the latest release
    unzip(paste0(path_data, file_details$name), overwrite = T, exdir = safe_path_data)
    
    # Create log record
    fileConn <- file(paste0(path_data, "download_log.txt"))
    writeLines(file_details$name, fileConn)
    close(fileConn)
    
    # Tidy up
    if (file.exists(paste0(path_data, file_details$name))){
      result <- file.remove(paste0(path_data, file_details$name))
    }
  }
  
  update_details <- web_get_update_details()
  
  file_details <- web_get_file_details(update_details$url)
  
  #Test whether local data is already latest release
  if (file.exists(paste0(path_data, "download_log.txt"))){
    
    # Read log record value
    rec <- read.table(paste0(path_data, "download_log.txt"), 
               colClasses = "character")[1, 1]
    
    # Download if log is not the most recent release
    if(rec==file_details$name){
      cat("Local GRID data cache already latest release\n")
    } else {
      web_get_file_download(file_details, path_data)
    }
    
  } else {
    # Download if no log found
    web_get_file_download(file_details, path_data)
  }
}
